### PR TITLE
[MKLDNN] Add LSTMP to v1.6.x

### PR DESCRIPTION
**Pull request by zixuanweeei**
## Description ##
This PR aims to support projection feature for LSTM towards the CPU platform.

### Essentials ###
- [x] Upgrade dnnl(mkldnn) to its v1.4 version
- [x] Change some statement of `mkldnn` to `dnnl`
- [x] Implement LSTMP from both native and DNNL pass
- [x] Add unit test for LSTMP
- [x] Add env var to control the initializing behavior of RNN
- [x] Remain input dim as int type in dnnl pass
- [x] Add logging message for MXNET_RNN_USE_WEIGHT_CACHE=1

## Comments ##
- This feature requires dnnl-v1.4rc porting to MXNet v1.6.x branch. Perhaps, there are several unexpected failures of unit tests. We need to have a test of CI in advance. And then switch to the formal release of dnnl-v1.4rc.
- Performance data will be added later.

@ciyongch @TaoLv @pengzhao-intel

### CODE REVIEW

**Comment by pengzhao-intel** [comment]
@sandeep-krishnamurthy @szha

**Comment by ciyongch** [comment]
Please note that all the changes in this PR are ready for review except a minor pending item: to upgrade mkldnn to v1.4rc version to finalize the PR.

File: src/operator/nn/mkldnn/mkldnn_rnn.cc [hunk 1]
```diff
@@ -470,66 +488,92 @@ inline void EmplaceNetArgs(mkldnn_args_map_t* net_args, const int arg_name,
...
-void MKLDNNRnnForward::SetWeightsMem(MKLDNNRnnMemMgr* mgr, void *w_ptr, void *b_ptr,
+void MKLDNNRnnForward::SetWeightsMem(void *w_ptr, void *b_ptr,
                                      const bool is_train, const int dtype) {
   using format_tag = mkldnn::memory::format_tag;
   auto mkldnn_dtype = get_mkldnn_type(dtype);
+  const size_t dtype_bytes = mshadow::mshadow_sizeof(dtype);
+
+  const size_t buffer_bytes = this->GetSize()  // byte number of the buffer
+      + (param_.workspace_size + param_.reserve_size) * dtype_bytes
+      + kMKLDNNAlign * 7;     // Add margin for alignment of seven times allocation for the
```
Comment (ciyongch) [inline src/operator/nn/mkldnn/mkldnn_rnn.cc:499]: Please remove the white-space to fix lint error.

File: cd/mxnet_lib/static/Jenkins_pipeline.groovy [hunk 2]
```diff
@@ -31,7 +31,7 @@ licenses = 'licenses/*'
 
 // libmxnet dependencies
 mx_deps = 'lib/libgfortran.so.3, lib/libquadmath.so.0'
-mx_mkldnn_deps = 'lib/libgfortran.so.3, lib/libquadmath.so.0, lib/libmkldnn.so.1, 3rdparty/mkldnn/build/install/include/mkldnn_version.h'
+mx_mkldnn_deps = 'lib/libgfortran.so.3, lib/libquadmath.so.0, lib/libdnnl.so.1, 3rdparty/mkldnn/build/install/include/mkldnn_version.h'
```
[the remark this answers is not in the document]
Comment (zixuanweeei) [reply cd/mxnet_lib/static/Jenkins_pipeline.groovy:34]: Thanks for the reminder of this.

**Comment by zixuanweeei** [comment]
@mxnet-bot run ci [windows-cpu]

**Comment by zixuanweeei** [comment]
@mxnet-bot run ci [centos-gpu]

File: src/operator/rnn_impl.h [hunk 3]
```diff
@@ -248,14 +258,26 @@ void LstmForwardInferenceSingleLayer(DType* ws,
...
       } else {
-        h[j][k] = ht;
         c[j][k] = ct;
       }
+      h[j][k] = ht;
+    }
+    if (P > 0) {
+      linalg_gemm(h, whr, r, alpha, beta, false, true);
+#pragma GCC diagnostic push
+#if __GNUC__ >= 8
```
Comment (TaoLv) [inline src/operator/rnn_impl.h:273]: Can you add comments for this check?

Comment (zixuanweeei) [reply src/operator/rnn_impl.h:273]: It was added by #17752 to fix cpu gcc8 -WError build. I forget to remove these lines. Maybe we can just drop them, right?

**Comment by sandeep-krishnamurthy** [comment]
@eric-haibin-lin @apeforest @ChaiBapchya - Can you please help review this PR?

**Comment by pengzhao-intel** [comment]
Seems there are problems in CI @zixuanweeei could you point out the failure cases and I think AWS colleagues can help double-check.

**Comment by zixuanweeei** [comment]
> Seems there are problems in CI @zixuanweeei could you point out the failure cases and I think AWS colleagues can help double-check.

I created #17964 to track the CI status for v1.6.x branch. Basically, we can see several failures which appear in both sides:

+ [unix-cpu] - R: CPU
+ [unix-cpu] - R: MKLDNN-CPU
+ [unix-gpu] - R: GPU
+ [unix-gpu] - Static build GPU 14.04 Python
+ [website] - R Docs

There are lots of failures with R. Besides, I believe `test_operator_gpu.test_parallel_random_seed_setting_for_context` is a flaky test (#15856).

@sandeep-krishnamurthy Could you help us to take a look for the CI failures?

**Comment by zixuanweeei** [comment]
@mxnet-bot run ci [windows-gpu]

**Comment by zixuanweeei** [comment]
After upgrading the MKLDNN (oneDNN) to the formal v1.4-rc, same failures appeared as mentioned above. I thinks it may be caused by the CI issues rather than the functionality issues of MKLDNN(oneDNN). Please take review @sandeep-krishnamurthy , thanks.

**Comment by sandeep-krishnamurthy** [comment]
@eric-haibin-lin - Can you take a look at this PR please?
@ChaiBapchya - Any pointers on why R tests are failing here?

**Comment by ChaiBapchya** [comment]
The issue related to R package test is fixed with importing the function first, using `library(methods)`
I've made the change to v1.6.x branch [testing it locally now]. 

Referring the doc for Reproducing the tests : https://cwiki.apache.org/confluence/display/MXNET/Reproducing+test+results

1. Requirements
```
pip3 install -r ci/requirements.txt --user
```
2. Build
```
ci/build.py --docker-registry mxnetci --platform ubuntu_cpu --docker-build-retries 3 --shm-size 500m /work/runtime_functions.sh build_ubuntu_cpu_openblas
```
3. Test
```
ci/build.py --docker-registry mxnetci --platform ubuntu_cpu --docker-build-retries 3 --shm-size 500m /work/runtime_functions.sh unittest_ubuntu_cpu_R
```

**Comment by aaronmarkham** [comment]
> The issue related to R package test is fixed with importing the function first, using `library(methods)`
> I've made the change to v1.6.x branch [testing it locally now].
> 
> Referring the doc for Reproducing the tests : https://cwiki.apache.org/confluence/display/MXNET/Reproducing+test+results
> 
> 1. Requirements
> 
> ```
> pip3 install -r ci/requirements.txt --user
> ```
> 
> 1. Build
> 
> ```
> ci/build.py --docker-registry mxnetci --platform ubuntu_cpu --docker-build-retries 3 --shm-size 500m /work/runtime_functions.sh build_ubuntu_cpu_openblas
> ```
> 
> 1. Test
> 
> ```
> ci/build.py --docker-registry mxnetci --platform ubuntu_cpu --docker-build-retries 3 --shm-size 500m /work/runtime_functions.sh unittest_ubuntu_cpu_R
> ```

R docs are broken on 1.6.x so maybe this fix can be applied to that pipeline too?
https://github.com/apache/incubator-mxnet/issues/17920

**Comment by zixuanweeei** [comment]
Thanks for your quick response @aaronmarkham @ChaiBapchya. Hope it will work soon. 

Besides, **[unix-gpu] - Static build GPU 14.04 Python** looks like being broken by some network issues. And it also appeared in #17964.

**Comment by zixuanweeei** [comment]
Performance comparison between the native and DNNL(oneDNN) implementation (tested on 28-core, 2.70GHz Intel CPU).

batch_size=64, sequence_len=50, input embedding_size=512
<table border=0 cellpadding=0 cellspacing=0 width=830 style='border-collapse:
 collapse;table-layout:fixed;width:623pt'>
 <col width=73 style='mso-width-source:userset;mso-width-alt:2346;width:55pt'>
 <col width=79 style='mso-width-source:userset;mso-width-alt:2517;width:59pt'>
 <col width=87 style='width:65pt'>
 <col width=68 style='mso-width-source:userset;mso-width-alt:2176;width:51pt'>
 <col width=97 style='mso-width-source:userset;mso-width-alt:3114;width:73pt'>
 <col width=121 style='mso-width-source:userset;mso-width-alt:3882;width:91pt'>
 <col width=97 style='mso-width-source:userset;mso-width-alt:3114;width:73pt'>
 <col width=121 style='mso-width-source:userset;mso-width-alt:3882;width:91pt'>
 <col width=87 style='width:65pt'>
 <tr height=21 style='height:16.0pt'>
  <td rowspan=2 height=42 class=xl67 width=73 style='border-bottom:.5pt solid black;
  height:32.0pt;width:55pt'>directions</td>
  <td rowspan=2 class=xl67 width=79 style='border-bottom:.5pt solid black;
  width:59pt'>layer_num</td>
  <td rowspan=2 class=xl67 width=87 style='border-bottom:.5pt solid black;
  width:65pt'>hidden_size</td>
  <td rowspan=2 class=xl67 width=68 style='border-bottom:.5pt solid black;
  width:51pt'>proj_size</td>
  <td colspan=2 class=xl69 width=218 style='border-left:none;width:164pt'>DNNL</td>
  <td colspan=2 class=xl69 width=218 style='border-left:none;width:164pt'>Native</td>
  <td rowspan=2 class=xl67 width=87 style='border-bottom:.5pt solid black;
  width:65pt'>Throughput Speedup</td>
 </tr>
 <tr height=21 style='height:16.0pt'>
  <td height=21 class=xl65 style='height:16.0pt;border-top:none;border-left:
  none'>fwd time</td>
  <td class=xl65 style='border-top:none;border-left:none'>fwd samples/sec</td>
  <td class=xl65 style='border-top:none;border-left:none'>fwd time</td>
  <td class=xl65 style='border-top:none;border-left:none'>fwd samples/sec</td>
 </tr>
 <tr height=21 style='height:16.0pt'>
  <td height=21 class=xl65 align=right style='height:16.0pt;border-top:none'>1</td>
  <td class=xl65 align=right style='border-top:none;border-left:none'>1</td>
  <td class=xl65 align=right style='border-top:none;border-left:none'>512</td>
  <td class=xl65 align=right style='border-top:none;border-left:none'>256</td>
  <td class=xl66 align=right style='border-top:none;border-left:none'>0.01</td>
  <td class=xl66 align=right style='border-top:none;border-left:none'>8312.23</td>
  <td class=xl66 align=right style='border-top:none;border-left:none'>0.04</td>
  <td class=xl66 align=right style='border-top:none;border-left:none'>1539.50</td>
  <td class=xl66 align=right style='border-top:none;border-left:none'>5.40</td>
 </tr>
 <tr height=21 style='height:16.0pt'>
  <td height=21 class=xl65 align=right style='height:16.0pt;border-top:none'>2</td>
  <td class=xl65 align=right style='border-top:none;border-left:none'>1</td>
  <td class=xl65 align=right style='border-top:none;border-left:none'>512</td>
  <td class=xl65 align=right style='border-top:none;border-left:none'>256</td>
  <td class=xl66 align=right style='border-top:none;border-left:none'>0.01</td>
  <td class=xl66 align=right style='border-top:none;border-left:none'>4416.66</td>
  <td class=xl66 align=right style='border-top:none;border-left:none'>0.06</td>
  <td class=xl66 align=right style='border-top:none;border-left:none'>1060.28</td>
  <td class=xl66 align=right style='border-top:none;border-left:none'>4.17</td>
 </tr>
 <tr height=21 style='height:16.0pt'>
  <td height=21 class=xl65 align=right style='height:16.0pt;border-top:none'>1</td>
  <td class=xl65 align=right style='border-top:none;border-left:none'>1</td>
  <td class=xl65 align=right style='border-top:none;border-left:none'>1024</td>
  <td class=xl65 align=right style='border-top:none;border-left:none'>512</td>
  <td class=xl66 align=right style='border-top:none;border-left:none'>0.02</td>
  <td class=xl66 align=right style='border-top:none;border-left:none'>3892.16</td>
  <td class=xl66 align=right style='border-top:none;border-left:none'>0.06</td>
  <td class=xl66 align=right style='border-top:none;border-left:none'>1048.26</td>
  <td class=xl66 align=right style='border-top:none;border-left:none'>3.71</td>
 </tr>
 <tr height=21 style='height:16.0pt'>
  <td height=21 class=xl65 align=right style='height:16.0pt;border-top:none'>2</td>
  <td class=xl65 align=right style='border-top:none;border-left:none'>1</td>
  <td class=xl65 align=right style='border-top:none;border-left:none'>1024</td>
  <td class=xl65 align=right style='border-top:none;border-left:none'>512</td>
  <td class=xl66 align=right style='border-top:none;border-left:none'>0.03</td>
  <td class=xl66 align=right style='border-top:none;border-left:none'>1979.32</td>
  <td class=xl66 align=right style='border-top:none;border-left:none'>0.10</td>
  <td class=xl66 align=right style='border-top:none;border-left:none'>626.18</td>
  <td class=xl66 align=right style='border-top:none;border-left:none'>3.16</td>
 </tr>
 <tr height=21 style='height:16.0pt'>
  <td height=21 class=xl65 align=right style='height:16.0pt;border-top:none'>1</td>
  <td class=xl65 align=right style='border-top:none;border-left:none'>3</td>
  <td class=xl65 align=right style='border-top:none;border-left:none'>512</td>
  <td class=xl65 align=right style='border-top:none;border-left:none'>256</td>
  <td class=xl66 align=right style='border-top:none;border-left:none'>0.02</td>
  <td class=xl66 align=right style='border-top:none;border-left:none'>3027.49</td>
  <td class=xl66 align=right style='border-top:none;border-left:none'>0.08</td>
  <td class=xl66 align=right style='border-top:none;border-left:none'>832.71</td>
  <td class=xl66 align=right style='border-top:none;border-left:none'>3.64</td>
 </tr>
 <tr height=21 style='height:16.0pt'>
  <td height=21 class=xl65 align=right style='height:16.0pt;border-top:none'>2</td>
  <td class=xl65 align=right style='border-top:none;border-left:none'>3</td>
  <td class=xl65 align=right style='border-top:none;border-left:none'>512</td>
  <td class=xl65 align=right style='border-top:none;border-left:none'>256</td>
  <td class=xl66 align=right style='border-top:none;border-left:none'>0.04</td>
  <td class=xl66 align=right style='border-top:none;border-left:none'>1537.22</td>
  <td class=xl66 align=right style='border-top:none;border-left:none'>0.18</td>
  <td class=xl66 align=right style='border-top:none;border-left:none'>361.05</td>
  <td class=xl66 align=right style='border-top:none;border-left:none'>4.26</td>
 </tr>
 <tr height=21 style='height:16.0pt'>
  <td height=21 class=xl65 align=right style='height:16.0pt;border-top:none'>1</td>
  <td class=xl65 align=right style='border-top:none;border-left:none'>3</td>
  <td class=xl65 align=right style='border-top:none;border-left:none'>1024</td>
  <td class=xl65 align=right style='border-top:none;border-left:none'>512</td>
  <td class=xl66 align=right style='border-top:none;border-left:none'>0.05</td>
  <td class=xl66 align=right style='border-top:none;border-left:none'>1346.44</td>
  <td class=xl66 align=right style='border-top:none;border-left:none'>0.18</td>
  <td class=xl66 align=right style='border-top:none;border-left:none'>349.79</td>
  <td class=xl66 align=right style='border-top:none;border-left:none'>3.85</td>
 </tr>
 <tr height=21 style='height:16.0pt'>
  <td height=21 class=xl65 align=right style='height:16.0pt;border-top:none'>2</td>
  <td class=xl65 align=right style='border-top:none;border-left:none'>3</td>
  <td class=xl65 align=right style='border-top:none;border-left:none'>1024</td>
  <td class=xl65 align=right style='border-top:none;border-left:none'>512</td>
  <td class=xl66 align=right style='border-top:none;border-left:none'>0.12</td>
  <td class=xl66 align=right style='border-top:none;border-left:none'>534.99</td>
  <td class=xl66 align=right style='border-top:none;border-left:none'>0.33</td>
  <td class=xl66 align=right style='border-top:none;border-left:none'>195.29</td>
  <td class=xl66 align=right style='border-top:none;border-left:none'>2.74</td>
 </tr>
 <![if supportMisalignedColumns]>
 <tr height=0 style='display:none'>
  <td width=73 style='width:55pt'></td>
  <td width=79 style='width:59pt'></td>
  <td width=87 style='width:65pt'></td>
  <td width=68 style='width:51pt'></td>
  <td width=97 style='width:73pt'></td>
  <td width=121 style='width:91pt'></td>
  <td width=97 style='width:73pt'></td>
  <td width=121 style='width:91pt'></td>
  <td width=87 style='width:65pt'></td>
 </tr>
 <![endif]>
</table>

**Comment by ciyongch** [comment]
@ChaiBapchya @aaronmarkham Thanks for looking into those failed R tests, I'm wondering whether https://github.com/apache/incubator-mxnet/issues/17920 can fix the CI failure or not?

**Comment by ChaiBapchya** [comment]
I've a PR #17993 it should fix the R related errors

**Comment by ciyongch** [comment]
> I've a PR #17993 it should fix the R related errors

Thanks @ChaiBapchya , hope this fix will be merged into v1.6.x soon to move this PR forward:)
Ping @eric-haibin-lin for a code review.

**Comment by ChaiBapchya** [comment]
Yes, waiting on the builds to pass and will get it merged soon!

**Comment by ChaiBapchya** [comment]
@zixuanweeei PR #17993 merged
Please rebase. Let's get this merged!
Thanks!

**Comment by ciyongch** [comment]
@ChaiBapchya Thanks for your prompt fix on the CI issues, now all the CI tests are passed:)

**Comment by ChaiBapchya** [comment]
Since this a new feature and it is not included in 1.6.0 pypi
should we cherry-pick this to v1.7.x in time for release of v1.7.0?
@zixuanweeei @ciyongch @pengzhao-intel wdyt?

**Comment by ciyongch** [comment]
I suggest to leave this new feature to the next release version (which requires upgrade oneDNN (ex: DNNL)), as the current version 1.7.0 is code freeze and suppose to only accept bug fixes, besides this feature is also included in master branch. The reason for applying it in 1.6.x branch is it's required by customer. 
@ChaiBapchya is there any special requirement for this feature in 1.7.0?

**Comment by ChaiBapchya** [comment]
@ciyongch While the customer request is resolved with 1.6.x backport, it made sense to include it in upcoming minor release. Backport to 1.6 should be included by default in 1.7 right?

**Comment by ciyongch** [comment]
>Backport to 1.6 should be included by default in 1.7 right?

Hi @ChaiBapchya , not sure if it's a common practice to do so. As there're several new features including (oneDNN) FP32 LSTMP/INT8 LSTM/INT8 LSTMP introduced in v1.6.x branch, are they all needed to go into 1.7.0 or only FP32 LSTMP is good enough? As currently INT8 LSTMP relies on non-release version of oneDNN which might be involved till it's matured and included in a formal release version. Previously MXNet is always pin to a formal release version of oneDNN.
cc @TaoLv to see if there's any other comments.
Thanks,
Ciyong.

**Comment by TaoLv** [comment]
@ChaiBapchya @ciyongch As far as I know, the feature lives on 1.6.x branch but is not released in any release yet. Given the code of 1.7.0 has been frozen for a while and to pay respect to the effort from release manager, I would suggest to not include this feature in the coming rc0. We can reconsider the request if we will have another rc1. The pined version of oneDNN is also a valid concern.

**Comment by ChaiBapchya** [comment]
@TaoLv Makes sense. Thanks for the clarification. Let's circle back to this point if rc1 is required.

**Comment by ciyongch** [comment]
Thanks for your understand and help making it clear @TaoLv @ChaiBapchya .

**Pull request merged**